### PR TITLE
Filter unsupported constructs in a separate pass

### PR DESCRIPTION
**Pull request by mcimadamore**
This PR moves all the logic to filter unsupported constructs in a dedicated visitor pass, hence removing all the checks from `OutputFactory`. This makes `OutputFactory` a lot simpler, as now this class only has to check for the presence of the `Skip` attribute.

One map in output factory (to keep track of generated structs) has been removed; we can in fact just set the `Skip` attribute on a scoped declaration as soon as we process it, to achieve the same effect. This map was also used in `visitTypedef`, so that we could fetch the correct supertype for the typedef declaration. This logic has now been replaced by making the `JavaName` attribute more expressive, so that we can deal with qualified names.

The other map could not be removed, and has been moved instead inside `NameMangler`. The issue is that we need some way to link a typedef type to the name of the functional interface associated with the typedef. Unfortunately, despite many attempts, I realized that the typedef type does not share anything in common with the typedef declaration; as such, piggy backing on existing attributes is not feasible. So for now, I kept a map, to retain existing behavior. It is possible that the IR might have to be tweaked to accommodate this use case.

Since `NameMangler` now correctly set the functional interface name on all variables whose type is either a function pointer, or a function pointer typedef, the logic in `OutputFactory` can, again, be simplified: now `OutputFactory` only needs to fetch the name attribute set by the mangler (if any).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (no review required)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jextract.git pull/151/head:pull/151` \
`$ git checkout pull/151`

Update a local copy of the PR: \
`$ git checkout pull/151` \
`$ git pull https://git.openjdk.org/jextract.git pull/151/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 151`

View PR using the GUI difftool: \
`$ git pr show -t 151`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jextract/pull/151.diff">https://git.openjdk.org/jextract/pull/151.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jextract/pull/151#issuecomment-1836669598)

### CODE REVIEW

File: src/main/java/org/openjdk/jextract/impl/OutputFactory.java [hunk 1]
```diff
@@ -232,50 +168,36 @@ public Void visitFunction(Declaration.Function funcTree, Declaration parent) {
...
-            return null;
-        }
 
         // check for function pointer type arguments
-        int i = 0;
         for (Declaration.Variable param : funcTree.parameters()) {
             Type.Function f = Utils.getAsFunctionPointer(param.type());
             if (f != null) {
-                if (! generateFunctionalInterface(param, f)) {
+                if (! generateFunctionalInterface(JavaFunctionalInterfaceName.getOrThrow(param), f)) {
```
Comment (JornVernee) [inline src/main/java/org/openjdk/jextract/impl/OutputFactory.java:176]: Pre-existing, but this looks potentially problematic: it seems like it is possible that we generate a FI class for one of the parameters, then find that a following parameter is Skipped, and drop the function. Leaving a dangling FI class for the other param (similar if the return type is skipped).

These FI classes are unique to the particular function AFAICS. See also my related comment in UnsupportedFilter. I think this part can be simplified more by putting the Skip on the function tree instead. Then we would never see an unsupported/Skipped type here.

File: src/main/java/org/openjdk/jextract/impl/UnsupportedFilter.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+        @Override
+        public String visitType(Type t, Void unused) {
+            return null;
+        }
+    };
+
+    static void warn(String msg) {
+        System.err.println("WARNING: " + msg);
+    }
```
Comment (JornVernee) [inline src/main/java/org/openjdk/jextract/impl/UnsupportedFilter.java:242]: Can be `private` as well I think?

File: src/main/java/org/openjdk/jextract/impl/NameMangler.java [hunk 3]
```diff
@@ -202,6 +224,11 @@ public Void visitVariable(Declaration.Variable variable, Declaration parent) {
         if (func != null) {
             String fiName = curScope.uniqueNestedClassName(variable.name());
             JavaFunctionalInterfaceName.with(variable, fiName);
+        } else if (variable.type() instanceof Delegated delegatedType) {
+            String typedefName = functionTypeDefNames.get(delegatedType.type());
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/main/java/org/openjdk/jextract/impl/NameMangler.java:228]: Yes, they are structurally equal (the old code also relied on this property)

File: src/main/java/org/openjdk/jextract/impl/UnsupportedFilter.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    public Void visitDeclaration(Declaration d, Declaration declaration) {
+        return null;
+    }
+
+    private void checkFunctionType(Declaration decl, Type.Function func) {
+        String unsupportedType = firstUnsupportedType(func);
+        if (unsupportedType != null) {
+            warn("skipping " + JavaName.getOrThrow(decl) + " because of unsupported type usage: " +
+                    unsupportedType);
+            Skip.with(func);
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/main/java/org/openjdk/jextract/impl/UnsupportedFilter.java]: Very good suggestion - thanks!

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**